Update gcc version to 12

jclarkeSTFC opened the pull request:
See #37519 for full details, but conda-forge use `gcc` version 12, so this updates our version to match. The developer environment and the conda build config have both been changed.

A few compiler warnings appeared, which I've also fixed:
- In `Chunker.cpp` there was an include missing.
- In `Quasi.cpp` there was an unnecessary (accidental) copy due to the specified type of the loop variable.
- In `InstrumentWidgetMaskTab.cpp` the local variable `parent` shadowed a method also called `parent()`, and it generated a warning about being unintialised.
- In `SpectrumDetectorMappingTest.h` a couple of variables were not initialised.
- In `SumOverlappingTubes.cpp` and `MuonAlgorithmHelper.cpp` I've reused the existing vector to prevent an `array-bounds` warning.
- Test folders in `Framework/` with warnings have the `-Wno-maybe-uninitialized` compiler flag to stop warnings coming from the use of `EXPECT_CALL` with the `_` arguments, which is a google test construction.
- Disabled a couple of warnings for `stdsoap2.cpp`.
- Qt test executables now have the `-Wno-maybe-uninitialized` compiler flag to disable the `maybe-uninitialized` warning.

See [here](https://gcc.gnu.org/onlinedocs/gcc/gcc-command-options/options-to-request-or-suppress-warnings.html) for information on the various `gcc` warnings.

Fixes #37519 

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

### To test (Linux-only):
- Remove your `mantid-developer` environment with `conda env remove -n mantid-developer`
- Delete `build/CMakeCache.txt`
- Create your `mantid-developer` environment with `mamba env create -f mantid-developer-linux.yml`
- Check that your environment has the right version by running `conda list | grep gxx`
- Run `cmake --preset=linux`, the compiler version detected should be 12.3 (or 12.* anyway)
- Compile
- Go and get a cup of tea
- Everything should compile and run as normal

Check that the Linux CI builds used version 12 instead of 10 by looking at the build logs, they'll need a clean build to pick up the compiler change ~(I'm waiting until after 6.10 is released before triggering those)~

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Update gcc version to 12

I also fixed some warnings, there was a missing include in Chunker.cpp,
an accidental cast + copy in Quasi.cpp, and an unitialised variable in
InstrumentWidgetMaskTab.cpp that also shadowed a method called parent().

jclarkeSTFC committed:
Added release notes

jclarkeSTFC committed:
Update cppcheck suppressions

jclarkeSTFC committed:
Fix compiler warnings in gcc 12

jclarkeSTFC commented:
There are a lot of warnings being generated in the Linux build, but not in a debug build. When using the `linux-64-ci` `cmake` preset the configuration is set to Release, and what seems to be happening is that when code from libraries flagged as system libraries is inlined, this can generate an error. In Debug that code won't generate an error because the header files are included using the `isystem` flag, which will suppress warnings from those files. But in Release that code is now inline and hence not in the system header. There's a similar discussion about this [here](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=109559), and the conclusion seemed to be that this is a feature not a bug. That thread also suggests this is a 'regression' from `gcc` version 12 onwards, which matches up with what's happening on this PR.

jclarkeSTFC committed:
Fix a couple of gcc warnings.

One is a -Wnonnull warning (CutMDTest), the other is less clear but
it's a -Warray-bounds warning, I'm hoping that this will fix it. It
looked like a vector that had been deleted was potentially being copied,
we'll see.

jclarkeSTFC committed:
Ignore some warnings in gcc 12

Also fixed a couple of cppcheck warnings and adjusted suppressions for
vms_convert.cpp that are now in the wrong place.

jclarkeSTFC committed:
Initialise this variable to prevent gcc warning

jclarkeSTFC committed:
Disable maybe-uninitialized warning for methods

jclarkeSTFC committed:
The default constructors here do not initialise

jclarkeSTFC committed:
Disable maybe-uninitialized warning for methods

jclarkeSTFC committed:
Ignore warning for test files

Using gcc 12 will generate a maybe-uninitialized warning when a test
file uses EXPECT_CALL from google mock with a _ (anything) argument

jclarkeSTFC committed:
Ignore maybe-uninitialized warnings

Ignoring warning in stdsoap2.cpp seems standard procedure, while
TypedPropertyValueHandler and IFunctionAdapter use boost::extract,
which seems to generate a maybe-uninitialized warning. MDGridBox looks
to be a mistake, I don't see how vertexCoord on line 493 is not
initialized.

jclarkeSTFC committed:
More boost::extract warnings

jclarkeSTFC committed:
Remove some array-bounds warnings from gcc 12

jclarkeSTFC committed:
Disable string-op overflow warning for gcc 12

jclarkeSTFC committed:
Use existing vector to prevent gcc warning

The old code generated an array-bounds warning in gcc 12, and it
doesn't seem a bad idea to use the existing vector instead of making
a new one.

jhaigh0 commented on docs/source/release/v6.11.0/Workbench/New_features/37519.rst:
```diff
@@ -0,0 +1 @@
+- Updated compiler on Linux to gcc version 12, which should improve performance in some circumstances
```
Should this be in `Framework/Dependancies`?

jclarkeSTFC replied:
Release note moved

jhaigh0 commented on Framework/API/test/SpectrumDetectorMappingTest.h:
```diff
@@ -116,8 +116,8 @@ class SpectrumDetectorMappingTest : public CxxTest::TestSuite {
 
   void test_array_constructor_null_inputs() {
-    specnum_t specs[2];
-    detid_t detids[2];
+    specnum_t specs[2] = {0, 1};
+    detid_t detids[2] = {0, 1};
     TS_ASSERT_THROWS(SpectrumDetectorMapping(nullptr, detids, 10), const std::invalid_argument &);
```
Does this still test for '`null_inputs`'?

jclarkeSTFC replied:
Yes the `nullptr` is the argument that it's checking to be null, and it uses either `specs` or `detids` for the other argument.

jhaigh0 replied:
right got it :+1:

jhaigh0 reviewed (commented):
Compiles with no additional warnings, had version 12.4 on my machine and on the CI build.

Wanted to check if all the surpressions are ones we can't fix / the compiler doesn't understand Mantid code very well? Or if its the case with cppcheck where its easier to surpress warnings now and fix them at a later date. Is it dangerous to surpress warnings for entire cmake lists?

jclarkeSTFC committed:
Move parent index check further up method

Get the parent index in one place in the method instead of duplicating
the code. Also moved the release note for this branch.

jclarkeSTFC commented on qt/widgets/instrumentview/src/InstrumentWidgetMaskTab.cpp:
```diff
@@ -476,7 +476,7 @@ void InstrumentWidgetMaskTab::singlePixelPicked(size_t pickID) {
   }
 
-  size_t parent;
+  size_t parentIndex = -1;
```
I've changed this

jclarkeSTFC commented:
I rebased due to a merge conflict in `Quasi.cpp`

robertapplin commented on qt/scientific_interfaces/ISISReflectometry/test/CMakeLists.txt:
```diff
@@ -1,3 +1,7 @@
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  add_compile_options(-Wno-maybe-uninitialized)
+endif()
+
```
I see this is repeated in a lot of CMakeLists, is it possible for this to go in a more common location?

Perhaps in here https://github.com/mantidproject/mantid/blob/main/buildconfig/CMake/CommonSetup.cmake or somewhere similar?

robertapplin replied:
Or perhaps this is the location for it? https://github.com/mantidproject/mantid/blob/fc864399055dc8c96200a84c1c521da3f96668aa/buildconfig/CMake/GNUSetup.cmake#L101

jclarkeSTFC replied:
I could do that but there would be a list of test folders specified in e.g. `GNUSetup.cmake`, so the build configuration for those test folders would then be split between their local `CMakeLists.txt` and the other file.

robertapplin replied:
Oh I see,  so we only want this compile option for these test libraries?

robertapplin replied:
In that case ignore me! Looks fine as it is :)

jclarkeSTFC replied:
Haha thanks, yes it's only some specific test folders that need the compiler flag.

robertapplin commented on Framework/TestHelpers/inc/MantidFrameworkTestHelpers/MDEventsTestHelper.h:
```diff
@@ -343,15 +343,15 @@ static MDGridBox<MDLeanEvent<nd>, nd> *makeMDGridBox(size_t split0 = 10, size_t
     splitter->setSplitInto(1, split1);
   // Set the size to 10.0 in all directions
-  auto box = std::make_unique<MDBox<MDLeanEvent<nd>, nd>>(splitter);
+  auto box = new MDBox<MDLeanEvent<nd>, nd>(splitter);
```
Just want to check, now that this is a raw pointer, does it get manually deleted by the MDGridBox?

jclarkeSTFC replied:
`MDGridBox` was previously getting passed a raw pointer, so that behaviour is the same, it's just that in the old code the `unique_ptr` would go out of scope when `makeMDGridBox` exited, leaving `MDGridBox` with a dangling pointer.

robertapplin replied:
Ah I see. Is it possible to create a unique pointer for `MDBox` and then std::move it into `MDGridBox` so the ownership is transferred to the `MDGridBox`? This way we won't need to worry about deleting the raw pointer when the `MDGridBox` is destructed

robertapplin replied:
The BoxController on line 338 is also a raw owning pointer, but as far as I can see it is also never deleted in the destructor for MDBox. Perhaps this might be a maintenance issue to start using unique_ptr's in this part of the code to make sure the lifetime of these objects are managed correctly. I'm happy to approve this if you think it should be solved as part of a separate issue?

jclarkeSTFC replied:
It looks like in all the tests that use `makeMDGridBox`, at the end of the test the `MDGridBox` and the `BoxController` are deleted manually, e.g. `MDGridBoxTest.h:300`, so yes I think this should be improved a bit, I can make the issue later.

I've changed the `makeMDGridBox` code back to how it was, but disabled the warning just for that method (hopefully I got the correct warning 😆).

jclarkeSTFC committed:
Disable warning for MDBox that's deleted

jclarkeSTFC commented:
I've rebased to fix a conflict in `conda/recipes/conda_build_config.yaml` caused by the update to clang being merged in.
